include/deal.II/differentiation/sd/symengine_tensor_operations.h

```diff
@@ -0,0 +1,196 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_differentiation_sd_symengine_tensor_operations_h
+#define dealii_differentiation_sd_symengine_tensor_operations_h
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_SYMENGINE
+
+#  include <deal.II/base/symmetric_tensor.h>
+#  include <deal.II/base/tensor.h>
+
+#  include <deal.II/differentiation/sd/symengine_number_types.h>
+#  include <deal.II/differentiation/sd/symengine_types.h>
+
+#  include <utility>
+#  include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Differentiation
+{
+  namespace SD
+  {
+    /**
+     * @name Symbolic variable creation
+     */
+    //@{
+
+    /**
+     * Return a vector of Expressions representing a vectorial symbolic
+     * variable with the identifier specified by @p symbol.
+     *
+     * For example, if the @p symbol is the string <tt>"v"</tt> then the vectorial
+     * symbolic variable that is returned represents the vector $v$. Each
+     * component of $v$ is prefixed by the given @p symbol, and has a suffix that
+     * indicates its component index.
+     *
+     * @tparam dim The dimension of the returned tensor.
+     * @param[in] symbol An indentifier (or name) to the vector of returned
+     *            symbolic variables.
+     * @return A vector (a rank-1 tensor) of symbolic variables with the name of
+     *         each individual component prefixed by @p symbol.
+     *
+     * @warning It is up to the user to ensure that there is no ambiguity in the
+     * symbols utilized within a section of code.
+     */
+    template <int dim>
+    Tensor<1, dim, Expression>
+    make_symbol_vector(const std::string &symbol);
```

Comment:
Oh, I love this! 👍

Reply:
Is there a reason why you always use `make_symbol_XXX` instead of `make_symbolic_XXX`?

Reply:
So, my thinking is the following:

It would be incorrect to call this `make_symbolic_vector`, because one could conceivably have a (scalar) object that symbolically represents an entire vector. This is sort of what we'd like to have done for our linear operators. So I'd like to reserve `make_symbolic_vector` for when non-commutative operations are supported so that we can actually have a scalar symbolic object representing any arbitrary vector. So, in general, I think that we should stick away from calling anything other than a scalar entity "symbolic".

So that left me with the verbose choices
- `make_symbol`
- `make_symbolic_function`
- `make_vector_of_symbols`
- `make_vector_of_symbolic_functions`
- `make_tensor_of_symbols`
- `make_tensor_of_symbolic_functions`
- `make_symmetric_tensor_of_symbols`
- `make_symmetric_tensor_of_symbolic_functions`

or, the shorthand I chose
- `make_symbol`
- `make_symbol_function`
- `make_symbol_vector`
- `make_symbol_function_vector`
- `make_symbol_tensor`
- `make_symbol_function_tensor`
- `make_symbol_symmetric_tensor`
- `make_symbol_function_symmetric_tensor`

This was primarily done out of laziness, so if you do prefer the first set of names or some alternative then I'd be happy to rename them.

Reply:
(Good question, btw :-)

Reply:
Ok, then I think the current naming is perfectly fine.

Reply:
I've given some thought to your criticism, and I think that it would be best if I do rename these functions to something more explicit of their intent/purpose. We have a number of utility functions with lengthy names in the library, and these wouldn't be called extremely often in a user's code. I think the pro's to a long name outweigh the con's.